### DISCUSSION TITLE
Row Range Selection

### DISCUSSION THREAD

**Discussion by mbofb** [mar10/wunderbaum#25] [Ideas]
Could in the grid part, a selection like in a spreadsheet, from a column and start row to the same column and end row be done - (and for me most important, copy & paste treating this range)

**Comment by mar10** [comment]
# Behavior in grids

One case could be: 
- click into a cell to select it
- then shift+click another cell to select a range

Another case:
- click into a cell to select it
- then ctrl+click another cell to select this cell in addition

Another case:
- click into a cell to select it
- then shift+cursor down to extend the range by one cell down

In general
- Clicking a cell *activates* the row and *selects* the cell exclusively (i.e. *deselects* all other cells if any)
  This cell becomes the **anchor** for further selections
- (?) Ctrl-clicking a selected cell removes it from the range
  The *anchor* remains as before

We should also handle the case of row-nav-mode (instead of cell-nav-mode) where cells are not separately selectable, but only whole rows (i.e. nodes).

We'd also need callbacks that prevent selection (e.g. we don't want to select cells in folder rows, or only multiple cell in the same column)

**Comment by mar10** [comment]
## Behavior in plain trees (or treegrids in row-nav mode)

Assume we have an *active* node and shift-click three rows below. Let's also assume that checkboxes are visible to reflect the *selected* state.

**Microsoft File Explorer and macOS Finder**

- Unlike Wunderbaum: There is no distinction between *active* and *selected*
  So clicking a node should *activate* and *select* it exclusively, i.e. deselect all other nodes if any.
  This node becomes the anchor for further selections
- Shift-clicking another node should 
  - **not** *activate* the new node. The previous node remains the anchor
  - *select* the new node and all nodes in-between
  - Callbacks or options could control if all or only some nodes should be accepted, e.g. based on their type or indentation level
- Ctrl-clicking another node should 
  - **not** *activate* the new node
  - toggle *selection* of the new node
- Simple clicks and cursor movements deselect all nodes and *activate* + *select* the new target node

**Alternative**

- We introduce a new state *marked* and treat it independently from *selected*.
- This would allow to *mark* a range of nodes and press <kbd>space</kbd> to toggle *selection* state or apply other bulk operations
- **Caveat** introducing a new independent state doubles the number of permutations that we need to visualize with CSS 
  (active, focused, hovered, selected, **marked**)
